[FEA] Add config argument to Qualification tool

cindyyuanjiang opened the pull request:
Fixes https://github.com/NVIDIA/spark-rapids-tools/issues/1847.

### Changes

This PR adds a config argument to Qualification tool to support user-provided qualx-conf.yaml files

### Usage

`spark_rapids qualification [...] --config <qualx-conf.yaml>`

leewyang reviewed (commented):
LGTM, aside from Ahmed's comments.  I think we should keep the `custom_model_file` for now (for backwards-compatibility and since it might be a fair amount of work).  And if we change the `config` arg here, we might want to change it in the other APIs (and docs) to be consistent.

sayedbilalbari commented on user_tools/src/spark_rapids_tools/cmdli/tools_cli.py:
```diff
@@ -50,6 +50,7 @@ def qualification(self,
                       submission_mode: str = None,
                       target_cluster_info: str = None,
                       tuning_configs: str = None,
+                      config: str = None,
```
My suggestion would be to use something like `qualx_config` or something that retains the context that this is a file related to the model or qualX.
The reason for the explicit mention is that tools takes a similar config file by the use of `tools_config`. Param names like `config`/`pipeline_config` make it a bit difficult to understand which config files impacts which part.

cindyyuanjiang replied:
thanks @bilalbari! `qualx_config` sounds reasonable to me, we can use this for `qualification`, `prediction`, and `train`, and for `train_and_evaluate`, we can name it `qualx_pipeline_config` to differentiate from the above.

cindyyuanjiang commented:
Updates in 2nd iteration:

- Update argument `config` to `qualx_config` in Qualification, Prediction, and Train CLI
- Updated argument `config` to `qualx_pipeline_config` in TrainAndEvaluate CLI
- Kept function argument `config` the same in `predict`, `train`, and `train_and_evaluate`
- Updated documentation

cindyyuanjiang commented on user_tools/docs/qualx.md:
````diff
@@ -152,10 +152,13 @@ Once satisfied with the model, just supply the path to this model in the `--cust
+
 #### Configuration
 The training and evaluation pipeline is configured via the `qualx-pipeline-conf.yaml` file, which should have the following format:
 ```yaml
````
@leewyang Do you think it's necessary to copy this yaml content into a file under resources and use it as an example in the `train_and_evaluate` `qualx_pipeline_config` argument description?

leewyang replied:
We could.  The main issue is that the `qualx_pipeline_config` is specific to a model/customer, e.g. eventlog locations, dataset names, etc, so it'd only be an example file (not a functional default).

cindyyuanjiang commented on user_tools/src/spark_rapids_tools/cmdli/tools_cli.py:
```diff
@@ -306,31 +312,31 @@ def train(self,
         Each invocation will create a new dataset JSON file with an incrementing number.
 
-        :param config: Path to YAML config file containing the required training parameters.
+        :param qualx_pipeline_config: Path to YAML config file containing the required training parameters.
```
@leewyang does this fallback to `resources/qualx-conf.yaml` if not provided by user?

cindyyuanjiang commented on user_tools/src/spark_rapids_pytools/rapids/qualification.py:
```diff
@@ -539,9 +539,11 @@ def __update_apps_with_prediction_info(self,
                                      output_info=output_info,
                                      model=estimation_model_args['customModelFile'],
+                                     config=estimation_model_args['config'],
                                      qual_handlers=[qual_handler])
```
updated

cindyyuanjiang commented on user_tools/docs/qualx.md:
```diff
@@ -152,10 +152,13 @@ Once satisfied with the model, just supply the path to this model in the `--cust
 spark_rapids train_and_evaluate \
---config qualx-pipeline-conf.yaml
+--qualx_pipeline_config qualx-pipeline-conf.yaml
 --ouptut_folder train_and_evaluate_output
```
fixed

cindyyuanjiang commented on user_tools/src/spark_rapids_pytools/rapids/qualx/train.py:
```diff
@@ -60,7 +60,7 @@ def _run_rapids_tool(self):
                 base_model=self.wrapper_options['base_model'],
                 features_csv_dir=self.wrapper_options['features_csv_dir'],
-                config=self.wrapper_options['config'],
+                config=self.wrapper_options['qualx_config'],
```
thanks @amahussein, I noticed this as well. It is not a typo. I was trying to follow the convention in the original files.

In `train.py`, the arguments are snake-case, while in `prediction.py`, the arguments are camel-case. (for `qualification.py` it is camel-case)

I am happy to make this consistent once we agree which case to use. cc: @leewyang

cindyyuanjiang replied:
Had an offline discussion, we will keep this since it is consistent with the original file. We could revisit this in the future to make the cases consistent everywhere.

cindyyuanjiang commented:
> > And if we change the `config` arg here, we might want to change it in the other APIs (and docs) to be consistent.
> 
> Thanks @leewyang @cindyyuanjiang I meant to change the `config` for the non-qualx CLIs. Especially, the qualX 'conf' argument of the Qual/Prof tools-CLI cmds should be self-explained that it is there to configure QualX. Otherwise, users will be confused between: autoTuner's conf, target cluster conf, qualx conf..and so on.

Thanks @amahussein, I see. I proposed my changes earlier: https://github.com/NVIDIA/spark-rapids-tools/pull/1848#discussion_r2263756569. I thought we are good to proceed from there.

I will revert the changes for `prediction` and `train` arguments.

cindyyuanjiang commented on user_tools/src/spark_rapids_tools/cmdli/argprocessor.py:
```diff
@@ -286,12 +286,14 @@ class EstimationModelArgProcessor(AbsToolUserArgModel):
             self.p_args['toolArgs']['estimationModel'] = QualEstimationModel.get_default()
         else:
             self.p_args['toolArgs']['estimationModel'] = self.estimation_model
+        self.p_args['toolArgs']['qualxConfig'] = self.qualx_config
```
Used camel-case here to be consistent with `estimationModel` before it

cindyyuanjiang commented on user_tools/docs/qualx.md:
```diff
@@ -152,10 +152,13 @@ Once satisfied with the model, just supply the path to this model in the `--cust
 spark_rapids train_and_evaluate \
---config qualx-pipeline-conf.yaml
---ouptut_folder train_and_evaluate_output
+--qualx_pipeline_config qualx-pipeline-conf.yaml
```
this is fixed, can you check again?

amahussein reviewed (approved):
LGTM!

Thanks @cindyyuanjiang
